File: src/bin/day9.cpp
#include <cmath>
#include <fmt/core.h>
#include <fstream>
#include <iostream>
#include <ostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <string_view>
#include <unordered_set>
#include <vector>

struct Coords {
  int row{};
  int col{};

  Coords() = default;

  Coords(int _row, int _col) : row(_row), col(_col) {}

  bool operator==(const Coords &other) const { return row == other.row && col == other.col; }

  Coords operator+(const Coords &other) const { return { row + other.row, col + other.col }; }

  Coords operator-(const Coords &other) const { return { row - other.row, col - other.col }; }

  Coords &operator+=(const Coords &other) {
    row += other.row;
    col += other.col;
    return *this;
  }

  static Coords normalized(const Coords &coord) {
    Coords other = coord;
    if (other.row != 0) { other.row = (other.row / std::abs(other.row)); }
    if (other.col != 0) { other.col = (other.col / std::abs(other.col)); }
    return other;
  }

  friend std::ostream &operator<<(std::ostream &out, const Coords &other) {
    out << fmt::format("Coords{{ row: {}, col: {} }}", other.row, other.col);
    return out;
  }
};

struct CoordHash {
  std::size_t operator()(const Coords &coord) const {
    return std::hash<int>()(coord.row) ^ std::hash<int>()(coord.col);
  }
};

struct Move {
  Coords coord;
  int mag;
};

Coords adjust_knot(const Coords &head, const Coords &tail) {
  const double dist = std::sqrt(std::pow(head.row - tail.row, 2) + std::pow(head.col - tail.col, 2));
  if (dist < 2.0) { return tail; }
  return tail + Coords::normalized(head - tail);
}

std::size_t knots(const std::vector<Move> &moves, std::size_t num_knots) {
  std::vector<Coords> tails(num_knots);
  std::unordered_set<Coords, CoordHash> visited;
  visited.insert(tails[0]);
  for (const auto &[offset, mag] : moves) {
    for (int _ = 0; _ < mag; ++_) {
      tails[0] += offset;
      for (std::size_t i = 1; i < num_knots; ++i) {
        tails[i] = adjust_knot(tails[i - 1], tails[i]);
        if (i == num_knots - 1) { visited.insert(tails[i]); }
      }
    }
  }
  return visited.size();
}

std::vector<Move> parse_input(const std::string &file_path) {
  std::ifstream ifs(file_path);
  std::string line;
  std::vector<Move> moves;
  while (std::getline(ifs, line)) {
    std::stringstream ss(line);
    char dir;
    int mag;
    ss >> dir >> mag;
    switch (dir) {
    case 'R':
      moves.push_back({ { 0, 1 }, mag });
      break;
    case 'L':
      moves.push_back({ { 0, -1 }, mag });
      break;
    case 'U':
      moves.push_back({ { 1, 0 }, mag });
      break;
    case 'D':
      moves.push_back({ { -1, 0 }, mag });
      break;
    default:
      throw std::runtime_error("invalid direction");
    }
  }
  return moves;
}

int main() {
  std::vector<Move> moves = parse_input("input.txt");
  auto part_one = knots(moves, 2);
  auto part_two = knots(moves, 10);
  fmt::print("part one: {}\n", part_one);
  fmt::print("part two: {}", part_two);
}

File: src/bin/day9.rs
use anyhow::Result;
use itertools::Itertools;
use std::collections::HashSet;

#[derive(Debug)]
struct Move {
    mag: u32,
    dir: (i32, i32),
}

#[derive(Debug, Hash, Eq, PartialEq, Clone, Copy, Default)]
struct Coords {
    row: i32,
    col: i32,
}

fn parse_moves(input: &str) -> Vec<Move> {
    input
        .lines()
        .map(|line| {
            let (dir, mag) = line.split_whitespace().collect_tuple().unwrap();
            let mag: u32 = mag.parse().unwrap();
            match dir {
                "R" => Move { mag, dir: (0, 1) },
                "L" => Move { mag, dir: (0, -1) },
                "U" => Move { mag, dir: (1, 0) },
                "D" => Move { mag, dir: (-1, 0) },
                _ => panic!("invalid direction"),
            }
        })
        .collect_vec()
}

fn tail_is_pulled(head_coords: &Coords, tail_coords: &Coords) -> bool {
    let dist = (((head_coords.row - tail_coords.row).pow(2)
        + (head_coords.col - tail_coords.col).pow(2)) as f64)
        .sqrt();
    dist.floor() as i32 >= 2
}

fn get_new_tail_coords(head_move: &Move, head_coords: &Coords) -> Coords {
    let (dr, dc) = head_move.dir;
    Coords {
        row: head_coords.row - dr,
        col: head_coords.col - dc,
    }
}

fn print_grid(knots: &[Coords]) {
    println!("{knots:?}");
    let mut grid = [['.'; 6]; 5];
    grid[0][0] = 's';
    for (idx, knot) in knots.iter().rev().enumerate() {
        let mut c = char::from_digit(idx as u32, 10).unwrap();
        if idx == 0 {
            c = 'H';
        }
        grid[knot.row as usize][knot.col as usize] = c;
    }
    for row in grid {
        println!("{row:?}");
    }
}

fn part_one(moves: &[Move]) -> u32 {
    let mut tail_coords = Coords::default();
    let mut head_coords = Coords::default();
    let mut ans = 1u32;
    let mut visited: HashSet<Coords> = HashSet::new();
    visited.insert(tail_coords);
    for m in moves {
        let (dr, dc) = m.dir;
        for _ in 0..m.mag {
            head_coords.row += dr;
            head_coords.col += dc;
            if tail_is_pulled(&head_coords, &tail_coords) {
                tail_coords = get_new_tail_coords(m, &head_coords);
                if !visited.contains(&tail_coords) {
                    visited.insert(tail_coords);
                    ans += 1;
                }
            }
        }
    }
    ans
}

fn part_two(moves: &[Move]) -> u32 {
    let mut knots = [Coords::default(); 10];
    let mut ans = 1u32;
    let mut visited: HashSet<Coords> = HashSet::new();
    visited.insert(knots[0]);
    for m in moves {
        let (dr, dc) = m.dir;
        for _ in 0..m.mag {
            knots[0].row += dr;
            knots[0].col += dc;
            for i in 1..=9 {
                if tail_is_pulled(&knots[i - 1], &knots[i]) {
                    knots[i] = get_new_tail_coords(m, &knots[i - 1]);
                    if i == 9 && !visited.contains(&knots[i]) {
                        visited.insert(knots[i]);
                        ans += 1;
                    }
                }
            }
            println!("{knots:?}\n");
        }
    }
    ans
}

fn main() -> Result<()> {
    let input = std::fs::read_to_string("sample.txt")?;
    let moves = parse_moves(&input);
    let part_one_ans = part_one(&moves);
    let part_two_ans = part_two(&moves);
    println!("part one: {part_one_ans}");
    println!("part two: {part_two_ans}");
    Ok(())
}
